svgnative/include/svgnative/SVGRenderer.h

```diff
@@ -104,20 +104,113 @@ using Paint = boost::variant<Color, Gradient>;
+ * for any possible two members of the set, all real numbers between the two
+ * must also be within the set. The word smallest here means that no other proper
+ * subset of that set will satisfy all the requirements.
+ * 5. The intersection of two intervals $A$ and $B$ is the one whose set is
+ * the intersection of the sets of $A$ and $B$.
+ */
+class Interval
+{
+  public:
+    Interval() = default;
+    Interval(float u): Interval(u, u) {}
+    Interval(float u, float v);
+
+    /**
+     * Returns the greatest possible real number that's small than every
+     * member of this interval's set.
+     */
+    float Min() { return a; }
+
+    /**
+     * Returns the smallest possible real number that's greater than every
+     * member of this interval's set.
+     */
+    float Max() { return b; }
+    operator bool() { return !IsEmpty(); }
+
+    /* Computes the intersection of this interval with another one */
+    Interval operator&(Interval other);
+
+    /* Computes the join of this interval with another one, think of an interval that
+     * contains both the intervals */
+    Interval operator|(Interval other);
+
+    /* Returns true if one interval contains another one */
+    bool Contains(Interval other);
+
+    /* Is the interval empty? having no points in its set? */
+    bool IsEmpty() { return a == b; }
+
+    float a = 0;
+    float b = 0;
+};
+
+/**
+ * Represents a rectangle.
+ *
+ * The following points hold true for a Rectangle:
+ * 1. A rectangle is simply defined by a pair of two Intervals. The set
+ * of a rectangle $R$ defined by two Intervals $A$ and $B$ is the set
+ * of all points $(x, y)$ such that $x$ is a member of Interval $A$'s
+ * set while $y$ is a member of Interval $B$'s set.
+ * 2. A rectangle is empty if its set is of length zero.
+ * 3. A rectangle $A$ contains a rectangle $B$ if all points of rectangle
+ * $B$'s set are also members of the set of rectangle $A$.
+ * 4. The intersection of two rectangles $A$ and $B$ is a rectangle whose
+ * set contains the elements that are common in the point set of $A$ and
+ * that of $B$.
+ * 5. The join of two rectangles $A$ and $B$ is a rectangle with the smallest
+ * possible point set such that it has all points of $A$ as well as those of
+ * set $B$ and is complete, meaning that for any two points in the set, all
+ * the points lying between will also be a part of the set. By smallest we mean
+ * that no other proper subset of that set should satisfy this requirement.
+ */
+class Rect
 {
+  public:
+    using IntervalPair = std::tuple<Interval, Interval>;
     Rect() = default;
-    Rect(float aX, float aY, float aWidth, float aHeight)
-        : x{aX}
-        , y{aY}
-        , width{aWidth}
-        , height{aHeight}
-    {
-    }
-    float x = std::numeric_limits<float>::quiet_NaN();
-    float y = std::numeric_limits<float>::quiet_NaN();
-    float width = std::numeric_limits<float>::quiet_NaN();
-    float height = std::numeric_limits<float>::quiet_NaN();
```

Comment:
We used these values on purpose to differ between actual empty rects and CTOR created rects, if you need {0, 0, 0, 0} I'd suggest to explicitly call the CTOR with 0s

Reply:
I'm doing that right now. But there can be serious problem with this. All of the rest of my functions such as `isEmpty`, `join`, `intersection` become buggy when it's an actual empty rect (as opposed to CTOR created one) as quiet_NaN doesn't work with equality or other operations I think. The programmer will have to exercise care too make sure never to create a `Rect` by doing `Rect{}`. Btw, why do you want to distinguish between the two? Or when would you create a `Rect` by doing `Rect{}`?